### PR TITLE
OBPIH-3442 Add preference type logic in PO

### PR DESCRIPTION
**Pull request by DominikaHatala**

### CODE REVIEW

File: grails-app/controllers/org/pih/warehouse/JsonController.groovy [hunk 1]
```diff
@@ -1738,25 +1740,33 @@ class JsonController {
...
-        productSuppliers = productSuppliers.collect {[
-            id: it.id,
-            code: it.code,
-            supplierCode: it.supplierCode,
-            text: it.code + ' ' + it.name,
-            manufacturerCode: it.manufacturerCode,
-            manufacturer: it.manufacturer?.id,
-        ]}
+        productSuppliers = productSuppliers.collect {
+            ProductSupplierPreference preference = it.productSupplierPreferences.find { it.destinationParty == destinationParty }
```
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/JsonController.groovy:1749]: Is there any reason not to move this into the query above. If not, ProductSupplier probably needs its own service at this point. So let's create ProductSupplierService and add a method to retrieve ProductSuppliers using the given criteria.

File: grails-app/controllers/org/pih/warehouse/order/OrderController.groovy [hunk 2]
```diff
@@ -623,6 +626,10 @@ class OrderController {
         Order order = Order.get(params.order.id)
         OrderItem orderItem = OrderItem.get(params.orderItem.id)
         ProductSupplier productSupplier = null
+        if (params.validationCode == 'BLOCK') {
```
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/order/OrderController.groovy:629]: Let's use the enum ValidationCode.BLOCK.

Comment (DominikaHatala) [reply grails-app/controllers/org/pih/warehouse/order/OrderController.groovy:629]: I'ts not working with ValidationCode.BLOCK

File: grails-app/controllers/org/pih/warehouse/JsonController.groovy (hunk 1 above)
Comment (DominikaHatala) [reply grails-app/controllers/org/pih/warehouse/JsonController.groovy:1749]: Which query do you mean? What we do above is checking if preference for this destination party exists and if it exists then we check if validation code is different than hide (and then we include this product source in select) or if preference for this destination party doesn't exist at all (and then we also include it). I'm not really sure if I know how would you like it to look like.

File: grails-app/controllers/org/pih/warehouse/order/OrderController.groovy (hunk 2 above)
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/order/OrderController.groovy:629]: Sorry you'd need to convert params.validationCode to ValidationCode enum. Something like this should work but you'd need to do some error handling for when params.validationCode is null.
```
ValidationCode validationCode = params.validationCode ? params.validationCode as ValidationCode : null
if (validationCode == ValidationCode.BLOCK) { 
```

Other options that might also work ... but I prefer the above approach
```
ValidationCode validationCode = params.validationCode as ValidationCode
if (params.validationCode == ValidationCode.BLOCK.toString()) { 
```
or 
```
ValidationCode validationCode = params.validationCode as ValidationCode
if (params.validationCode == ValidationCode.BLOCK.name()) { 
```

File: grails-app/controllers/org/pih/warehouse/JsonController.groovy (hunk 1 above)
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/JsonController.groovy:1749]: Yeah I guess I was wondering if we could make it work so you can match on preferences in the query.
```
def productSuppliers = ProductSupplier.createCriteria().list { 
  eq("product", product)
  eq("supplier", suppler)
  productSupplierPreferences { 
    and { 
       eq("destinationParty", destinationParty)
       not { 
          eq("validationCode", ValidationCode.HIDE)
       }
    }
}
```
}

Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/JsonController.groovy:1749]: You might need to include something like this 
```
(productSupplierPreferences is empty or ...)
```

**Comment by DominikaHatala** [comment]
@jmiranda PR is updated.

File: grails-app/controllers/org/pih/warehouse/order/OrderController.groovy [hunk 3]
```diff
@@ -21,12 +21,15 @@ import org.pih.warehouse.core.Document
...
 import org.pih.warehouse.core.UomService
 import org.pih.warehouse.core.User
+import org.pih.warehouse.core.ValidationCode
 import org.pih.warehouse.product.Product
 import org.pih.warehouse.product.ProductSupplier
 import org.pih.warehouse.shipping.Shipment
 import org.pih.warehouse.shipping.ShipmentItem
 import org.springframework.web.multipart.MultipartFile
 
+import javax.xml.bind.ValidationException
```
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/order/OrderController.groovy:31]: Not the right ValidationException. It should be a grails class.

Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/order/OrderController.groovy:31]: Doesn't look like this is being used so you can probably get rid of the import.

**Review by jmiranda — CHANGES_REQUESTED** [review]
LGTM other than the issue I brought up below. Once that is fixed feel free to merge.

**Review by awalkowiak — APPROVED** [review]

**Pull request merged**